### DISCUSSION TITLE
per fragment properties

### DISCUSSION THREAD

**Discussion by kuitpo** [BoothGroup/Vayesta#90] [Q&A]
While I gather Vayesta is aimed at speeding calculation of whole system properties rather than calculating fragment properties, I wonder if it is possible to obtain fragment properties such as energy and dipole? Ideally the sum of the fragment energies would sum to the whole system energy, but energy of a fragment and the energy of its interaction with its environment would also have its uses.

Ta.

**Comment by maxnus** [comment]
Hi Ta,

In fact, the total correlation energy (from the standard "wave function functional") is calculated as a linear sum over fragment contributions. These can be found under `fragment.results.e_corr`. If you also want to partition the mean-field energy, there is a 
function `fragment.get_fragment_mf_energy()` [here](https://github.com/BoothGroup/Vayesta/blob/be227313c17c10d183e3dc9c7a522099cc426b6a/vayesta/core/qemb/fragment.py#LL385C31-L385C31).

The problem with "per fragment properties" is that they do not represent observables. This starts from the very beginning, from the ambiguous question of how electron density can be assigned to a fragment of a system. It then appears again, when splitting the interaction energy between fragments. In Vayesta, we do this splitting using ambiguous projections all the time, but as we only perform the splitting to apply individual, local approximations and look at the sum of the split quantity in the end, we can simply see it as a technique to speed up the calculation. It is possible that fragment quantities can be predictive, but this would need to be supported by data.

**Reply by kuitpo** [reply]
Hi maxnus,

Okay, back to Vayesta for a while.
Summing the fragment energies doesn't give me anything close to the total. 
"we=-1437.6" is more than double the total.
What am I doing wrong?
The CCSD D1 statistic is labelled unsatisfactory, but would that give rise to such an error?

```
mol = gto.M(verbose=4, basis='cc-pVDZ', atom=atoms, \
              max_memory=30000)
mf = pyscf.scf.RHF(mol)
mf.kernel()
emb = vayesta.ewf.EWF(mf, solver='CCSD', bath_options=dict(threshold=1e-6), solver_options=dict(init_guess='CISD'))
with emb.iaopao_fragmentation() as f:
  f.add_atomic_fragment([1,5,4,6,7,8,0,9,10,11])
  f.add_atomic_fragment([3,2,12,16])
  f.add_atomic_fragment([13,17,18])
  f.add_atomic_fragment([15,14,19,23])
  f.add_atomic_fragment([20,24,22,25,26,27,21,28,29,30])
emb.kernel()

we = 0.0
for i, frag in enumerate(emb.get_fragments()):
  ce = frag.results.e_corr
  mfe = frag.get_fragment_mf_energy()
  te = ce + mfe
  te_kc = te * 627.509608
  print(f'frag {i} ce={ce:.8f} mfe={mfe:.8f} te={te:.8f} te_kc={te_kc:.8f}')
  we += te
print(f'we={we:.8f}')

print("E(HF)=        %+16.8f Ha" % mf.e_tot)
print("E(Emb. CCSD)= %+16.8f Ha" % emb.e_tot)
```
output:
frag 0 ce=-0.48753144 mfe=-304.64289540 te=-305.13042683 te_kc=-191472.27453093
frag 1 ce=-0.48202677 mfe=-361.22611493 te=-361.70814170 te_kc=-226975.33421039
frag 2 ce=-0.15000890 mfe=-109.27879484 te=-109.42880374 te_kc=-68667.62574171
frag 3 ce=-0.48284956 mfe=-359.17003896 te=-359.65288852 te_kc=-225685.64308999
frag 4 ce=-0.48192401 mfe=-301.21879238 te=-301.70071639 te_kc=-189320.09827376
we=-1437.62097719
E(HF)=           -610.01230739 Ha
E(Emb. CCSD)=    -612.09664806 Ha

**Comment by ghb24** [comment]
Just to add to this, local non-energetic expectation values can be obtained from N-representable RDMs once reconstructed from the fragmented wave functions (noting arbitrariness as discussed by Max's point above re. choice of projector). Examples of this can be found in e.g. `examples/ewf/molecules/31-population-analysis.py` or `32-static-corrfuncs.py`.

**Reply by kuitpo** [reply]
Thanks, will look into all of that.

It is well known that fragments are not observable, much as for atoms and bonds in molecules. Fragments arising from appealing theory may be quite useful in rationalizing chemistry, just as bonds arising from a theory of bonding may be. One property a fragment analysis must have to be plausible (on the road to being appealing) is for changes in fragment properties to be smooth as the molecule structure or environment changes smoothly. Here fragment analyses routinely get into trouble immediately. If orbitals have integer occupations, orbital assignment to fragments changing as the molecule structure or environment changes gives major discontinuities in fragment properties. At least one fragment analysis provides a means avoiding discontinuity by keeping orbital to fragment assignments constant by nominating one molecular structure/environment as model for assignments that the rest of the set of structures also use (incidentally, the speed of that analysis happens to be quite limiting of molecular size). A more appealing approach is  
https://pubs.aip.org/aip/jcp/article/156/22/224113/2841375/Split-electrons-in-partition-density-functional
While this is only applied to diatoms so far (and is a DFT), it does address the problem I mention above. It also has an appealing principle for fragments - they are of equal electronegativity.

Is there presently some feature of Vayesta that may be used to avoid this orbital assignment discontinuity problem without nominating orbitals? For larger systems and number of fragments, referring to orbitals is not practical - need to refer to only atom identities in defining fragments.

It is desirable that a fragment analysis have these characteristics:
1. The total of the energies of the analysis is very close to equal the total system energy given by a non-fragment analysis 
2. Does not suffer the discontinuity problem mentioned above
3. Has an appealing principle for fragments
4. Can specify fragments using atom identity only without regard to manner of bonding between the atoms or reference to orbitals
5. Speed

Vayesta is doing very well on speed (5 - its main objective), and perhaps 1 and 4. 2 and 3 would make it useful for analysis.

**Reply by maxnus** [reply]
Perhaps the following remarks are useful:

We usually use "intrinsic atomic orbitals" (https://pubs.acs.org/doi/10.1021/ct400687b) to define fragments in a subspace (which nonetheless spans the entire occupied space of the mean-field reference). IAOs are not generally integer-occupied, and neither are the resulting fragments. Integer occupation seems prohibitively restrictive if the aim is to assign orbitals to "atoms", see e..g H_2^+.

Since IAOs span the occupied space entirely, they can be used to fragment mean-field (HF or DFT) quantities, such as the energy, directly.
In Vayesta, we also use IAOs to fragment the correlation energy of reference-dependent post-SCF methods, such as CCSD.
You can even do this for FCI, though it becomes less appealing, since the fragmentation is reference-dependent, whereas the underlying method is not. And finally, IAOs will vary as smoothly as the underlying mean-field does, if there is discontinuous behaviour in the mean-field, we can also expect to see it in the IAOs.

**Comment by maxnus** [comment]
I think you just need to add the classical core-core repulsion energy
(`mf.energy_nuc()` if memory serves).

Even if the CCSD is not satisfactory for the system, the fragmented
energies should still add up correctly.
